### ISSUE TITLE
xywhr in OBB result have changed

### ISSUE THREAD

**Issue by MarcelloCuoghi** [ultralytics/ultralytics#19642]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

Recently I updated my environment to the latest version of ultralytics and found that the xywhr result for the OBB task is somehow broken. I was using it to get the centre position of each detection, the width and height and the rotation. Before the update I always had the width as the largest dimension and the height as the smaller, and the rotation relative to the width.
After the update, I found that sometimes the width is the smaller dimension and the height is the larger one, so the rotation is also skewed by 90 degrees.
Going back to older version fix the problem.

### Environment

Ultralytics 8.3.87 🚀 Python-3.12.7 torch-2.6.0+cpu CPU (Intel Core(TM) i7-10850H 2.70GHz)
Setup complete ✅ (12 CPUs, 31.7 GB RAM, 609.9/952.5 GB disk)

OS                  Windows-11-10.0.22631-SP0
Environment         Windows
Python              3.12.7
Install             git
Path                C:\Users\my_user\Desktop\yolo_test\.venv\Lib\site-packages\ultralytics
RAM                 31.73 GB
Disk                609.9/952.5 GB
CPU                 Intel Core(TM) i7-10850H 2.70GHz
CPU count           12
GPU                 None
GPU count           None
CUDA                None

numpy               ✅ 2.1.1<=2.1.1,>=1.23.0
matplotlib          ✅ 3.10.1>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 11.1.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.15.2>=1.4.1
torch               ✅ 2.6.0>=1.8.0
torch               ✅ 2.6.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.21.0>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 7.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.14>=2.0.0

### Minimal Reproducible Example

Input image
![Image](https://github.com/user-attachments/assets/7c378fa6-7015-4cff-8c2c-e52420ed5fce)

# Ultralitics 8.3.87
With ultralytics package version 8.3.87 run this code:

```
from pathlib import Path
import torch
from ultralytics import YOLO

model_file = Path("yolo11x-obb.pt")
model = YOLO(model_file, task="OBB")

img = Path("./images/P0006.png")

results = model.predict(
    source=img,
    conf=0.8,
    imgsz=640,
    device=torch.cuda.current_device() if torch.cuda.device_count() > 0 else "CPU",
)

for res in results:
    for i, pts in enumerate(res.obb):
        if int(pts.xywhr[0][2]) < int(pts.xywhr[0][3]):
            print(
                f"{i})\tcenter x {int(pts.xywhr[0][0])}\tcenter y {int(pts.xywhr[0][1])}\twidth {int(pts.xywhr[0][2])}\theight {int(pts.xywhr[0][3])}\trotation {pts.xywhr[0][4]}"
            )
```
## Output:
```
# 3)      center x 322    center y 796    width 24        height 104      rotation 0.7707133889198303
# 9)      center x 329    center y 428    width 25        height 96       rotation 0.8094407916069031
# 11)     center x 325    center y 470    width 24        height 95       rotation 0.8357727527618408
# 12)     center x 324    center y 835    width 25        height 96       rotation 0.7769593000411987
# 16)     center x 329    center y 387    width 26        height 96       rotation 0.7910935282707214
# 17)     center x 586    center y 361    width 24        height 102      rotation 0.8455972075462341
# 19)     center x 582    center y 1246   width 24        height 93       rotation 0.8130463361740112
# 20)     center x 332    center y 345    width 25        height 95       rotation 0.8040628433227539
# 21)     center x 593    center y 394    width 25        height 104      rotation 0.8151286244392395
# 22)     center x 575    center y 1212   width 24        height 101      rotation 0.808992326259613
# 23)     center x 325    center y 674    width 24        height 96       rotation 0.8091808557510376
# 24)     center x 592    center y 673    width 25        height 102      rotation 0.8232788443565369
# 25)     center x 591    center y 798    width 25        height 104      rotation 0.8026906847953796
# 35)     center x 321    center y 878    width 25        height 93       rotation 0.787294864654541
# 40)     center x 585    center y 763    width 24        height 103      rotation 0.7789315581321716
# 41)     center x 590    center y 839    width 25        height 102      rotation 0.8089277148246765
# 44)     center x 318    center y 920    width 26        height 103      rotation 0.8125594854354858
# 47)     center x 592    center y 435    width 24        height 101      rotation 0.8140760660171509
# 50)     center x 325    center y 512    width 25        height 100      rotation 0.8281809091567993
# 51)     center x 312    center y 964    width 25        height 97       rotation 0.7953415513038635
# 54)     center x 333    center y 1306   width 24        height 99       rotation 0.854654848575592
# 55)     center x 317    center y 758    width 25        height 101      rotation 0.7970958948135376
# 60)     center x 318    center y 1119   width 25        height 104      rotation 0.8262746930122375
# 66)     center x 329    center y 708    width 24        height 94       rotation 0.8186021447181702
# 68)     center x 316    center y 1160   width 24        height 102      rotation 0.8195444345474243
# 69)     center x 583    center y 1286   width 24        height 100      rotation 0.8252146244049072
# 77)     center x 585    center y 722    width 24        height 103      rotation 0.7846349477767944
# 79)     center x 315    center y 1200   width 24        height 101      rotation 0.8189680576324463
# 81)     center x 147    center y 1238   width 21        height 58       rotation 1.5592427253723145
# 82)     center x 857    center y 1059   width 22        height 58       rotation 1.5689480304718018
# 84)     center x 312    center y 1282   width 24        height 98       rotation 0.8073341846466064
# 88)     center x 316    center y 1240   width 24        height 99       rotation 0.802503764629364
# 89)     center x 313    center y 1004   width 26        height 105      rotation 0.8169794082641602
# 90)     center x 582    center y 1163   width 23        height 98       rotation 0.8214647173881531
# 91)     center x 148    center y 1295   width 22        height 60       rotation 1.564757227897644
# 92)     center x 850    center y 1001   width 21        height 52       rotation 1.564372181892395
# 95)     center x 588    center y 881    width 26        height 103      rotation 0.7998103499412537
# 96)     center x 586    center y 1041   width 24        height 103      rotation 0.8159875273704529
# 98)     center x 853    center y 1029   width 23        height 60       rotation 1.5595061779022217
# 100)    center x 587    center y 959    width 25        height 103      rotation 0.8381613492965698
# 103)    center x 583    center y 1084   width 23        height 101      rotation 0.8027064204216003
# 107)    center x 603    center y 1310   width 24        height 101      rotation 0.8306283354759216
# 108)    center x 589    center y 920    width 26        height 104      rotation 0.8245478272438049
# 109)    center x 586    center y 1001   width 24        height 101      rotation 0.8359718322753906
# 110)    center x 319    center y 1038   width 24        height 102      rotation 0.8157163858413696
# 113)    center x 161    center y 364    width 22        height 66       rotation 1.5607739686965942
# 114)    center x 851    center y 1088   width 22        height 61       rotation 1.5662956237792969
# 115)    center x 581    center y 1124   width 23        height 101      rotation 0.8175578713417053
# 116)    center x 316    center y 1080   width 25        height 103      rotation 0.8084856271743774
```

# Ultralitics 8.3.32

## yolo checks
Ultralytics 8.3.32 🚀 Python-3.12.7 torch-2.6.0+cpu CPU (Intel Core(TM) i7-10850H 2.70GHz)
Setup complete ✅ (12 CPUs, 31.7 GB RAM, 613.0/952.5 GB disk)

OS                  Windows-11-10.0.22631-SP0
Environment         Windows
Python              3.12.7
Install             git
RAM                 31.73 GB
Disk                613.0/952.5 GB
CPU                 Intel Core(TM) i7-10850H 2.70GHz
CPU count           12
GPU                 None
GPU count           None
CUDA                None

numpy               ✅ 2.1.1>=1.23.0
matplotlib          ✅ 3.10.1>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 11.1.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.15.2>=1.4.1
torch               ✅ 2.6.0>=1.8.0
torchvision         ✅ 0.21.0>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 7.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.14>=2.0.0
numpy               ✅ 2.1.1<2.0.0; sys_platform == "darwin"
torch               ✅ 2.6.0!=2.4.0,>=1.8.0; sys_platform == "win32"

Running same code as above no output, so no results with width < height as expected.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @MarcelloCuoghi, thank you for bringing this to our attention 🚀! We appreciate your detailed report and the effort you put into sharing the context, code, and outputs. 

We recommend checking out our [Docs](https://docs.ultralytics.com/) for additional resources, including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, which may help clarify or address related behavior.

If this is indeed a 🐛 Bug Report, your provided information is helpful, but to assist us further in debugging, could you double-check that your code is a valid [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/)? An MRE simplifies the issue to its core, ensuring we can reproduce it quickly on our end.

### Steps to Proceed:
1. **Upgrade:** Make sure you're using the latest version of the `ultralytics` package. Upgrade in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) by running:
   ```bash
   pip install -U ultralytics
   ```

2. **Verify Environment:** Ensure your environment dependencies match the [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml). YOLO can run on several supported environments:
   - **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
   - **Google Cloud** Deep Learning VM: See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
   - **Amazon** Deep Learning AMI: See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
   - **Docker Image:** See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

3. **Testing:** If the issue persists, try running your code in a different verified environment (e.g., the free GPU Notebooks above) to confirm if the behavior is environment-specific.

### Community Resources:
Join our Ultralytics community for additional support and discussions:
- Real-time chat: [Discord](https://discord.com/invite/ultralytics) 🎧
- In-depth discussions: [Discourse](https://community.ultralytics.com/)
- Community sharing: [Subreddit](https://reddit.com/r/Ultralytics)

### CI Status:
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>  
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing, ensuring correct operation of YOLO across tasks and environments.

This is an automated response to guide you in resolving the issue 🤖. An Ultralytics engineer will review this further and assist you soon. Thank you for your patience and contribution to improving Ultralytics! 🚀

**Comment by Y-T-G** [comment]
It was updated so that the degrees are always less than 90 degrees to be consistent with what the model uses for training. The OBB uses the OpenCV definition, where the angles are always less than 90 degrees. The width isn't guaranteed to be larger than height under OpenCV definition.

If you want, you can manually swap the height and width and offset by 90.

**Comment by Y-T-G** [comment]
The angle is with respect to the positive x-axis, measured clockwise. So I am not sure why you find it skewed. It shouldn't be skewed if you're correctly using the angle.

**Comment by MarcelloCuoghi** [comment]
Ok, I understand.
The problem about the angle is that with respect to the previous results, but having swapped the width and height of the box is correct, so now in my application I also check the w and h of each box and eventually swap them and rotate myself the angle so that always the w has the greater value.

Thank you for your explanation

**Comment by glenn-jocher** [comment]
Thank you for sharing your solution! You're correct that the OpenCV convention used in our OBB implementation doesn't guarantee width > height, which is why we regularize angles to [0, π/2) radians. This ensures compatibility with standard computer vision workflows. For reference, the OBB format documentation details this behavior and conversion methods. We appreciate you adapting your application accordingly and welcome any future questions about YOLO implementations! 🚀

**Comment by urm9ril** [comment]
@glenn-jocher Can you post a link to the document?
